sql/ballots/ballots.go

```diff
@@ -213,62 +207,80 @@ func LatestLayer(db sql.Executor) (types.LayerID, error) {
 	return lid, nil
 }
 
+// FirstInEpoch returns the first ballot referencing the specified ATX in the epoch.
+// NOTE: it does not mark the ballot as malicious, if the smesher that published the ATX is!
 func FirstInEpoch(db sql.Executor, atx types.ATXID, epoch types.EpochID) (*types.Ballot, error) {
-	return inEpoch(db, atx, epoch, "asc")
+	var (
+		ballot *types.Ballot
+		dErr   error
+	)
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, atx.Bytes())
+		stmt.BindInt64(2, int64(epoch.FirstLayer()))
+		stmt.BindInt64(3, int64((epoch+1).FirstLayer()-1))
+	}
+	dec := func(stmt *sql.Statement) bool {
+		var bid types.BallotID
+		stmt.ColumnBytes(0, bid[:])
+		ballot, dErr = decodeBallot(bid, stmt.ColumnReader(1))
+		if dErr != nil {
+			return false
+		}
+		return ballot.EpochData == nil
+	}
+	rows, err := db.Exec(`
+		select id, ballot from ballots
+		where atx = ?1 and layer between ?2 and ?3
+		order by layer asc limit 1
+	`, enc, dec)
+	if err != nil {
+		return nil, fmt.Errorf("ballot by atx %s: %w", atx, err)
+	}
+	if dErr != nil {
+		return nil, fmt.Errorf("decode ballot by atx %s: %w", atx, err)
+	}
+	if rows == 0 {
+		return nil, sql.ErrNotFound
+	}
+	return ballot, err
 }
 
+// LastInEpoch returns the last ballot referencing the specified ATX in the epoch.
+// NOTE: it does not mark the ballot as malicious, if the smesher that published the ATX is!
 func LastInEpoch(db sql.Executor, atx types.ATXID, epoch types.EpochID) (*types.Ballot, error) {
-	return inEpoch(db, atx, epoch, "desc")
-}
-
-func inEpoch(db sql.Executor, atx types.ATXID, epoch types.EpochID, order string) (*types.Ballot, error) {
 	var (
-		bid     types.BallotID
-		ballot  types.Ballot
-		nodeID  types.NodeID
-		rows, n int
-		err     error
+		ballot *types.Ballot
+		dErr   error
 	)
 	enc := func(stmt *sql.Statement) {
 		stmt.BindBytes(1, atx.Bytes())
 		stmt.BindInt64(2, int64(epoch.FirstLayer()))
 		stmt.BindInt64(3, int64((epoch+1).FirstLayer()-1))
 	}
 	dec := func(stmt *sql.Statement) bool {
+		var bid types.BallotID
 		stmt.ColumnBytes(0, bid[:])
-		stmt.ColumnBytes(1, nodeID[:])
-		if n, err = codec.DecodeFrom(stmt.ColumnReader(2), &ballot); err != nil {
-			if err != io.EOF {
-				err = fmt.Errorf("ballot by atx %s: %w", atx, err)
-				return false
-			}
-		} else if n == 0 {
-			err = fmt.Errorf("ballot by atx missing data %s", atx)
+		ballot, dErr = decodeBallot(bid, stmt.ColumnReader(1))
+		if dErr != nil {
 			return false
 		}
-		ballot.SetID(bid)
-		ballot.SmesherID = nodeID
-		if stmt.ColumnInt(3) > 0 {
-			ballot.SetMalicious()
-		}
-		// only ref ballot has valid EpochData
-		if ballot.EpochData != nil {
-			return false
-		}
-		return true
+		return ballot.EpochData == nil
```

Comment:
Shouldn't it always return `false`? `LastInEpoch` returns only one ballot. Btw, it probably doesn't matter anyway because of the `limit 1` in the query...